File: drestaurant-libs/drestaurant-customer/src/test/java/com/drestaurant/customer/domain/CustomerAggregateTest.java
package com.drestaurant.customer.domain;

import java.math.BigDecimal;

import org.axonframework.messaging.interceptors.BeanValidationInterceptor;
import org.axonframework.messaging.interceptors.JSR303ViolationException;
import org.axonframework.test.AxonAssertionError;
import org.axonframework.test.aggregate.AggregateTestFixture;
import org.axonframework.test.aggregate.FixtureConfiguration;
import org.junit.Before;
import org.junit.Test;

import com.drestaurant.common.domain.model.AuditEntry;
import com.drestaurant.common.domain.model.Money;
import com.drestaurant.common.domain.model.PersonName;
import com.drestaurant.customer.domain.api.CreateCustomerCommand;
import com.drestaurant.customer.domain.api.CustomerCreatedEvent;

public class CustomerAggregateTest {

	private FixtureConfiguration<Customer> fixture;
	private AuditEntry auditEntry;
	private AuditEntry auditEntry2;
	private static final String WHO = "johndoe";
	private Money orderLimit = new Money(BigDecimal.valueOf(1000000));

	@Before
	public void setUp() throws Exception {
		fixture = new AggregateTestFixture<>(Customer.class);
		fixture.registerCommandDispatchInterceptor(new BeanValidationInterceptor<>());
		auditEntry = new AuditEntry(WHO);
		auditEntry2 = new AuditEntry(WHO + "2");
	}

	@Test
	public void createCustomerTest() throws Exception {
		PersonName name = new PersonName("Ivan", "Dugalic");
		CreateCustomerCommand createCustomerCommand = new CreateCustomerCommand(name ,orderLimit, auditEntry);
		CustomerCreatedEvent customerCreatedEvent = new CustomerCreatedEvent(name, orderLimit, createCustomerCommand.getTargetAggregateIdentifier(), auditEntry);

		fixture
		.given()
		.when(createCustomerCommand)
		.expectEvents(customerCreatedEvent);
	}

	@Test(expected = AxonAssertionError.class)
	public void createCustomerAxonAssertionErrorTest() throws Exception {
		PersonName name = new PersonName("Ivan", "Dugalic");
		CreateCustomerCommand createCustomerCommand = new CreateCustomerCommand(name, orderLimit, auditEntry);
		// Setting the WRONG auditEntry
		CustomerCreatedEvent customerCreatedEvent = new CustomerCreatedEvent(name, orderLimit, createCustomerCommand.getTargetAggregateIdentifier(), auditEntry2);

		fixture
		.given()
		.when(createCustomerCommand)
		.expectException(AxonAssertionError.class);

	}

	@Test(expected = JSR303ViolationException.class)
	public void createCustomerJSR303ViolationTest() throws Exception {
		// Setting NULL for a name
		PersonName name = new PersonName(null, "Dugalic");
		CreateCustomerCommand createCustomerCommand = new CreateCustomerCommand(name, orderLimit, auditEntry);
		CustomerCreatedEvent customerCreatedEvent = new CustomerCreatedEvent(name, orderLimit, createCustomerCommand.getTargetAggregateIdentifier(), auditEntry);

		fixture
		.given()
		.when(createCustomerCommand)
		.expectException(JSR303ViolationException.class);
	}

}

File: drestaurant-libs/drestaurant-customer/src/test/kotlin/com/drestaurant/customer/domain/CustomerAggregateTest.kt
package com.drestaurant.customer.domain

import com.drestaurant.common.domain.api.model.AuditEntry
import com.drestaurant.common.domain.api.model.Money
import com.drestaurant.common.domain.api.model.PersonName
import com.drestaurant.customer.domain.api.CreateCustomerCommand
import com.drestaurant.customer.domain.api.CreateCustomerOrderCommand
import com.drestaurant.customer.domain.api.CustomerCreatedEvent
import com.drestaurant.customer.domain.api.CustomerOrderCreatedEvent
import com.drestaurant.customer.domain.api.model.CustomerId
import org.axonframework.messaging.interceptors.BeanValidationInterceptor
import org.axonframework.test.AxonAssertionError
import org.axonframework.test.aggregate.AggregateTestFixture
import org.axonframework.test.aggregate.FixtureConfiguration
import org.junit.Before
import org.junit.Test
import java.math.BigDecimal
import java.util.*

class CustomerAggregateTest {

    private lateinit var fixture: FixtureConfiguration<Customer>
    private val who = "johndoe"
    private val auditEntry: AuditEntry = AuditEntry(who, Calendar.getInstance().time)
    private val auditEntry2: AuditEntry = AuditEntry(who + "2", Calendar.getInstance().time)
    private val orderLimit = Money(BigDecimal.valueOf(1000000))

    @Before
    fun setUp() {
        fixture = AggregateTestFixture(Customer::class.java)
        fixture.registerCommandDispatchInterceptor(BeanValidationInterceptor())
    }

    @Test
    fun createCustomerTest() {
        val name = PersonName("Ivan", "Dugalic")
        val createCustomerCommand = CreateCustomerCommand(name, orderLimit, auditEntry)
        val customerCreatedEvent = CustomerCreatedEvent(name, orderLimit, createCustomerCommand.targetAggregateIdentifier, auditEntry)

        fixture
                .given()
                .`when`(createCustomerCommand)
                .expectEvents(customerCreatedEvent)
    }

    @Test(expected = AxonAssertionError::class)
    fun createCustomerAxonAssertionErrorTest() {
        val name = PersonName("Ivan", "Dugalic")
        val createCustomerCommand = CreateCustomerCommand(name, orderLimit, auditEntry)
        // Setting the WRONG auditEntry
        val customerCreatedEvent = CustomerCreatedEvent(name, orderLimit, createCustomerCommand.targetAggregateIdentifier, auditEntry2)

        fixture
                .given()
                .`when`(createCustomerCommand)
                .expectException(AxonAssertionError::class.java)

    }

    @Test
    fun createCustomerOrderTest() {
        val name = PersonName("Ivan", "Dugalic")
        val createCustomerOrderCommand = CreateCustomerOrderCommand(CustomerId("customerId"), orderLimit, auditEntry)
        val customerCreatedEvent = CustomerCreatedEvent(name, orderLimit.add(Money(BigDecimal.ONE)), createCustomerOrderCommand.targetAggregateIdentifier, auditEntry)
        val customerOrderCreatedEvent = CustomerOrderCreatedEvent(orderLimit, CustomerId("customerId"), createCustomerOrderCommand.customerOrderId, auditEntry)
        fixture
                .given(customerCreatedEvent)
                .`when`(createCustomerOrderCommand)
                .expectEvents(customerOrderCreatedEvent)
    }

    @Test
    fun createCustomerOrderFailOrderLimitTest() {
        val name = PersonName("Ivan", "Dugalic")
        val createCustomerOrderCommand = CreateCustomerOrderCommand(CustomerId("customerId"), orderLimit, auditEntry)
        val customerCreatedEvent = CustomerCreatedEvent(name, orderLimit, createCustomerOrderCommand.targetAggregateIdentifier, auditEntry)
        val customerOrderCreatedEvent = CustomerOrderCreatedEvent(orderLimit, CustomerId("customerId"), createCustomerOrderCommand.customerOrderId, auditEntry)
        fixture
                .given(customerCreatedEvent)
                .`when`(createCustomerOrderCommand)
                .expectException(UnsupportedOperationException::class.java)
    }

}
